### ISSUE TITLE
Internal Viewer crashes when scrolling down via Fn+Down on macOS

### ISSUE THREAD

**Issue by lprot** [doublecmd/doublecmd#1680]
### Version

Double Commander
Version: 1.2.0 alpha
Revision: 11332
Commit: afc3a9fc9
Build date: 2024/05/20
Lazarus: 3.99.0.0
FPC: 3.2.2
Platform: x86_64-Darwin-cocoa
OS version: Mac OS X 12.7.5

### Description

To reproduce, unpack provided zip file. Press F3 on 0.boot.img then press 7 times Fn+Down. DC crashes.
If it does not crash, then just keep holding Fn+Down and it crashes later while scrolling pages quickly.

[0.boot.zip](https://github.com/doublecmd/doublecmd/files/15447791/0.boot.zip)

**Comment by lprot** [comment]
@alexx2000 stil reproducible on Double Commander
Version: 1.2.0 alpha
Revision: 11348
Commit: a2da252b1
Build date: 2024/05/28
Lazarus: 3.99.0.0
FPC: 3.2.2
Platform: x86_64-Darwin-cocoa
OS version: Mac OS X 12.7.5

**Comment by rich2014** [comment]
it's not a Lazarus or LCL/Cocoa bug, but FPC.

in `Iconvert()` of unit `iconvenc_dyn`, which convert encoding by calling the glibc function `iconvres`.

`iconvres` behaves differently on different platforms, and `Iconvert()` doesn't handle that correctly.

mainly the following:

```pascal
if lerr = ESysEILSEQ then // unknown char, skip
  begin
    Dst^ := Src^;
    Inc(Src);
    Inc(Dst);
    Dec(InLen);
    Dec(OutLen);     <--- may be zero here on macOS!
  end
```

**Comment by rich2014** [comment]
@lprot 

fixed by 01c8c0c, test the snapshoot please.

**Comment by lprot** [comment]
@rich2014 tested and cannot reproduce it on Double Commander
Version: 1.2.0 alpha
Revision: 11352
Commit: 01c8c0c81
Build date: 2024/06/01
Lazarus: 3.99.0.0
FPC: 3.2.2
Platform: x86_64-Darwin-cocoa
OS version: Mac OS X 12.7.5

Thank you!
